### ISSUE TITLE
[Bug] eksctl create nodegroup counts nodegroups as created+failed

### ISSUE THREAD

**Issue by matti** [weaveworks/eksctl#4970]
### What were you trying to accomplish?

Create a node group and then do things on it

```
while true do;
  eksctl create nodegroup -f my-ng.yaml && break
done

aws autoscaling create-or-update-tags my-ng ... # fails
```

### What happened?

eksctl create nodegroup create was unsuccessful due to aws internal error, but it still exited with 0 causing my scrip to go forward.

workaround:

```
while true; do
  eksctl create nodegroup -f my-ng.yaml
  
  eksctl get nodegroup --region="$REGION" --cluster="$CLUSTER_NAME" my-ng && break
done
```

**Versions**
```
eksctl 0.87.0
```

**Comment by Skarlso** [comment]
Can you post the logs please to see if it detected that there were any errors and this is maybe just a cobra thing?

**Comment by Skarlso** [comment]
There should be at least something like this:

```
failed to create nodegroups for cluster...
```

**Comment by Skarlso** [comment]
Can't reproduce:

```
➜  test-override-with-containerd eksctl create nodegroup --cluster test-override-10 --name test-ng-2 --nodes 1
2022-03-18 10:20:38 [ℹ]  eksctl version 0.89.0-dev+24f3d481.2022-03-17T22:15:09Z
2022-03-18 10:20:38 [ℹ]  using region us-west-2
2022-03-18 10:20:39 [ℹ]  will use version 1.21 for new nodegroup(s) based on control plane version
2022-03-18 10:20:44 [ℹ]  nodegroup "test-ng-2" will use "" [AmazonLinux2/1.21]
2022-03-18 10:20:46 [ℹ]  1 existing nodegroup(s) (ng-1) will be excluded
2022-03-18 10:20:46 [ℹ]  1 nodegroup (test-ng-2) was included (based on the include/exclude rules)
2022-03-18 10:20:46 [ℹ]  will create a CloudFormation stack for each of 1 managed nodegroups in cluster "test-override-10"
2022-03-18 10:20:46 [ℹ]
2 sequential tasks: { fix cluster compatibility, 1 task: { 1 task: { create managed nodegroup "test-ng-2" } }
}
2022-03-18 10:20:46 [ℹ]  checking cluster stack for missing resources
2022-03-18 10:20:47 [ℹ]  cluster stack has all required resources
2022-03-18 10:20:48 [ℹ]  building managed nodegroup stack "eksctl-test-override-10-nodegroup-test-ng-2"
2022-03-18 10:20:49 [ℹ]  deploying stack "eksctl-test-override-10-nodegroup-test-ng-2"
2022-03-18 10:20:49 [ℹ]  waiting for CloudFormation stack "eksctl-test-override-10-nodegroup-test-ng-2"
2022-03-18 10:21:04 [ℹ]  waiting for CloudFormation stack "eksctl-test-override-10-nodegroup-test-ng-2"
2022-03-18 10:21:22 [ℹ]  waiting for CloudFormation stack "eksctl-test-override-10-nodegroup-test-ng-2"
2022-03-18 10:21:42 [ℹ]  waiting for CloudFormation stack "eksctl-test-override-10-nodegroup-test-ng-2"
2022-03-18 10:21:43 [✖]  unexpected status "DELETE_IN_PROGRESS" while waiting for CloudFormation stack "eksctl-test-override-10-nodegroup-test-ng-2"
2022-03-18 10:21:43 [ℹ]  fetching stack events in attempt to troubleshoot the root cause of the failure
2022-03-18 10:21:43 [✖]  AWS::EKS::Nodegroup/ManagedNodeGroup: CREATE_FAILED – "Resource creation cancelled"
2022-03-18 10:21:43 [!]  AWS::CloudFormation::Stack/eksctl-test-override-10-nodegroup-test-ng-2: DELETE_IN_PROGRESS – "User Initiated"
2022-03-18 10:21:43 [ℹ]  1 error(s) occurred and nodegroups haven't been created properly, you may wish to check CloudFormation console
2022-03-18 10:21:43 [ℹ]  to cleanup resources, run 'eksctl delete nodegroup --region=us-west-2 --cluster=test-override-10 --name=<name>' for each of the failed nodegroup
2022-03-18 10:21:43 [✖]  waiting for CloudFormation stack "eksctl-test-override-10-nodegroup-test-ng-2": ResourceNotReady: failed waiting for successful resource state
Error: failed to create nodegroups for cluster "test-override-10"
➜  test-override-with-containerd echo $?
1
➜  test-override-with-containerd
```

**Comment by matti** [comment]
sorry I don't have logs anymore but this has happened 3 times already (I'm constantly running these up and down)

The failed run looks like this in cloudformation console:
<img width="1460" alt="image" src="https://user-images.githubusercontent.com/3893/158976870-4e6d9ddd-6a4d-4b16-a3a1-36d106f0c39c.png">

**Comment by Skarlso** [comment]
Interesting. 🤔 

I did try it, you can see the output above. It looks like, I got a return code of 1 and it exited with an internal cloud formation error.

Anything more? Can you share the config file?

**Comment by matti** [comment]
so it happened again on 2 clusters - my retry logic made it eventually succeed.

```
waiting for CloudFormation stack "eksctl-test-1-nodegroup-g-1-t-2-4-pre-p9-1-2022-03-18-10-50-24"
unexpected status "ROLLBACK_IN_PROGRESS" while waiting for CloudFormation stack "eksctl-test-1-nodegroup-g-1-t-2-4-pre-p9-1-2022-03-18-10-50-24"
fetching stack events in attempt to troubleshoot the root cause of the failure
AWS::EKS::Nodegroup/ManagedNodeGroup: DELETE_IN_PROGRESS
AWS::IAM::Policy/PolicyAWSLoadBalancerController: DELETE_IN_PROGRESS
AWS::IAM::Policy/PolicyEBS: DELETE_IN_PROGRESS
AWS::IAM::Policy/PolicyAutoScaling: DELETE_IN_PROGRESS
AWS::EKS::Nodegroup/ManagedNodeGroup: CREATE_FAILED – "Resource handler returned message: \"[Issue(Code=NodeCreationFailure, Message=Instances failed to join the kubernetes cluster, ResourceIds=[i-0aa418edfcfb5145b])] (Service: null, Status Code: 0, Request ID: null, Extended Request ID: null)\" (RequestToken: 8d21f799-cd07-7dc8-527f-b855aaae22ec, HandlerErrorCode: GeneralServiceException)"
 1 error(s) occurred and nodegroups haven't been created properly, you may wish to check CloudFormation console
to cleanup resources, run 'eksctl delete nodegroup --region=eu-north-1 --cluster=test-1 --name=<name>' for each of the failed nodegroup
waiting for CloudFormation stack "eksctl-test-1-nodegroup-g-1-t-2-4-pre-p9-1-2022-03-18-10-50-24": ResourceNotReady: failed waiting for successful resource state
Error: failed to create nodegroups for cluster "test-1"
```

^-- this happened 2x

here's the nodegroup yaml

```
apiVersion: eksctl.io/v1alpha5
kind: ClusterConfig

metadata:
  name: 'test-1'
  region: 'eu-north-1'

managedNodeGroups:
  - name: 'g-1-t-2-4-pre-p9-1-2022-03-18-10-50-24'
    instancePrefix: 'g-1'
    instanceName: 'g-1-t-2-4-pre-p9-1-2022-03-18-10-50-24'
    labels:
      test: 'yes'
    volumeType: 'gp3'
    volumeSize: 8
    volumeIOPS: 3000
    volumeThroughput: 125
    minSize: 0
    maxSize: 1
    spot: true
    iam:
      withAddonPolicies:
        autoScaler: true
        imageBuilder: true
        ebs: true
        albIngress: true
    instanceSelector:
      vCPUs: 2
      memory: '4'

    ami: ami-0f47e499c06cba4fc
    preBootstrapCommands:
      - date > /tmp/eksler.start
      - echo 'changing root password'
      - echo 'root:badpassword' | sudo chpasswd
      - echo 'password changed'
      - sudo yum install -y htop iotop nano screen bind-utils
      - sudo amazon-linux-extras install epel -y
      - sudo yum install -y stress
      - date > /tmp/eksler.done
    overrideBootstrapCommand: |
      #!/bin/bash
      date > /tmp/eksler.bootstrap.start
      /etc/eks/bootstrap.sh test-1 |& tee /tmp/eksler.bootstrap.output
      echo 0 > /tmp/eksler.bootstrap.status
      date > /tmp/eksler.bootstrap.done
```

*but* I have 5 of nodegroup creates running concurrently (to speed things up)

I'll add status code output printing in my wrapper next just to be sure, but since I have `set -e` in the bash script I don't think the script would have continued if eksctl create would have returned status 1 instead of 0.


I'll spend the rest of the day anyway creating clusters/nodegroups anyway so this will happen again. I'll share more logs etc later.

**Comment by matti** [comment]
and all my code is here: https://github.com/matti/eksler

```
git clone https://github.com/matti/eksler
cd eksler
# asdf install / ensure that versions in .tool_versions are in-place
bin/test-main         eu-north-1  test-0          ipv6  1 1.21
```

^-- this creates cluster "test-0" in eu-north-1, using ipv6, 1.21 kuberenetes and creates 7 nodegroups in parallel: https://github.com/matti/eksler/blob/78dd5d4a18b683df68515dc540d326b2a841eca6/bin/test-main#L20

**Comment by matti** [comment]
added debug print https://github.com/matti/eksler/commit/a905ad327fb348fc22d069bdc11ab76d9baf9840

**Comment by Skarlso** [comment]
Thanks Matti for all the things. :) I'll look into it.

I don't know how that could have happened. You see above that eksctl does return 1 on a failure. There is no way ( at least none I can see in the code right now ) that it would not `return err` in case of an error which is the only way of it exiting with a none error status code. 🤔 

I'll take a tour again and will try to reproduce something.

**Comment by Skarlso** [comment]
```
  eksctl create nodegroup -f my-ng.yaml && break
```

BTW that exists if the exit code is 0. Did you meant to write `||` here?

**Comment by matti** [comment]
>BTW that exists if the exit code is 0. Did you meant to write || here?
whops, yes - anyway in my testing/scripting it is correct

**Comment by Skarlso** [comment]
@matti Sorry, but, didn't you say your script uses the right thing? But I can see a `&& break` there. :)

**Comment by Skarlso** [comment]
Oh never mind. That's intentional so it runs as longs as it doesn't succeed.

**Comment by matti** [comment]
yeah, thats why I am pretty sure that eksctl exits with 0

**Comment by Skarlso** [comment]
Let's see what this outputs: 
```
              set +e
                eksctl --verbose $EKSCTL_LOG_LEVEL create nodegroup -f /tmp/eksler.$REGION.$CLUSTER_NAME.nodegroup.$nodegroup.yml
                echo "CREATE EXITED with: $?"
              set -e
```

I would be surprised if it's an error and it still says 0. But you never know. :) I might be surprised. :))

**Comment by matti** [comment]
so:

>AWS::EKS::Nodegroup/ManagedNodeGroup: CREATE_FAILED – "Resource handler returned message: \"[Issue(Code=NodeCreationFailure, Message=Instances failed to join the kubernetes cluster, ResourceIds=[i-00521e6fb963853bf])] (Service: null, Status Code: 0, Request ID: null, Extended Request ID: null)\" (RequestToken: 378e75a7-0279-4aa6-6a3e-35a25227f0ac, HandlerErrorCode: GeneralServiceException)"

happens and then

>waiting for CloudFormation stack "eksctl-test-1-nodegroup-g-1-t-16-32-pre-p7-1-2022-03-18-18-55-02": > ResourceNotReady: failed waiting for successful resource state
>Error: failed to create nodegroups for cluster "test-1"

and
>CREATE EXITED with: 1

**BUT**

then my script tries to get the nodegroup and it gets

>Error: ResourceNotFoundException: No node group found for name: g-1-t-16-32-pre-p7-1-2022-03-18-18-55-02

which is as expected.

Then my script tries to create it again, but from the output I can see that

>12 existing nodegroup(s) (... g-1-t-16-32-pre-p7-1-2022-03-18-18-55-02,...

so that it is "already created" even when it just failed

and then, since `eksctl nodegroup create` "does not create anything", eg it says

> created 0 managed nodegroup(s) in cluster "test-1"

then it exits with 0 - which might be wrong because the command is `eksctl nodegroup create` - not `ensure` (or `apply`) - IMO any "create" should exit with error if it doesn't create anything. 

so TL;DR - if eksctl nodegroup create fails it exits with 1, but then create which doesn't create also exits with 0 causing confusion.

from where does eksctl fetch this "12 existing nodegroups" because that nodegroup just failed + rollbacked, so it can not be existing ?

**Comment by matti** [comment]
In my cluster there are 6 nodegroups, but eksctl says "12 existing nodegroup(s)" - it's listing the failed stacks also (!) tried to search the eksctl code for "existing nodegroup(s)" but didn't find any :o

**Comment by matti** [comment]
workarounding this now like this: https://github.com/matti/eksler/blob/fbea28334199b1661b9742a147f5066172825fd6/bin/eksler#L378

**Comment by Skarlso** [comment]
>IMO any "create" should exit with error if it doesn't create anything.

Hmm, hmmmm, depends. If it didn't need to create anything then why should it error? Though I can understand an argument that that would be apply and not create. But there are certainly scenarios where create without error but no result is fine.

>code for "existing nodegroup(s)" but didn't find any :o

Because it's generated from

```go
	// Remote ones will be excluded
	if f.remoteNodegroups.Len() > 0 {
		logger.Info("%d existing %s(s) (%s) will be excluded", f.remoteNodegroups.Len(), "nodegroup", strings.Join(f.remoteNodegroups.List(), ","))
	}
```

Sneaky ;)

In any case if there are no other issues regarding exit code, can we close this ticket? :) If you have a different issue, please open a separate one. Thanks! :)

**Comment by matti** [comment]
>can we close this ticket? :)

I just try again with more precise description as I still think this create failed and exit 0 on retry is a bug: #4977 but that is just semantic wrestling and this is why I've written my eksler wrapper to change/fix behavior for my liking.

However, this is the real bug: https://github.com/weaveworks/eksctl/issues/4978

now we can close this ticket :)
